File: app/src/main/java/com/example/gymbuddy/view/AddWorkoutActivity.java
package com.example.gymbuddy.view;

import android.animation.ObjectAnimator;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.Spinner;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.example.gymbuddy.R;
import com.example.gymbuddy.helpers.Utils;
import com.example.gymbuddy.helpers.Helpers;
import com.example.gymbuddy.model.Workout;
import com.google.android.material.textfield.TextInputLayout;

import java.util.ArrayList;

public class AddWorkoutActivity extends AppCompatActivity {

    private EditText edtWorkoutName,edtMuscleGroup,edtMuscleGroupSecondary;
    private Spinner spinnerWorkoutType;
    private TextInputLayout tilWorkoutName, tilMuscleGroup, tilMuscleGroupSecondary;
    private Button btnAdd;
    private ImageView imgWorkoutName;
    boolean isEdited = false, changeMade = false;
    ArrayList<Workout> allWorkouts;
    int position;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        setTheme(R.style.Theme_GymBuddy);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_add_workout);

        Helpers.setupActionBar(getString(R.string.addNewWorkout),"",getSupportActionBar(),this);

        edtWorkoutName = findViewById(R.id.edtWorkoutName);
        spinnerWorkoutType = findViewById(R.id.spinnerWorkoutType);
        edtMuscleGroup = findViewById(R.id.edtMuscleGroup);
        edtMuscleGroupSecondary = findViewById(R.id.edtMuscleGroupSecondary);
        tilMuscleGroupSecondary=findViewById(R.id.tilMuscleGroupSecondary);
        btnAdd = findViewById(R.id.btnAdd);
        imgWorkoutName = findViewById(R.id.imgWorkoutName);

        edtWorkoutName.requestFocus();
        InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
        imm.toggleSoftInput(InputMethodManager.SHOW_FORCED, 0);

//        edtMuscleGroupSecondary.setEnabled(false);
        tilMuscleGroupSecondary.setEnabled(false);
//        edtMuscleGroupSecondary.getBackground().setColorFilter(Color.GRAY, PorterDuff.Mode.SRC_IN);

        edtMuscleGroup.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
//                  edtMuscleGroupSecondary.setEnabled(s.length() != 0);
                  tilMuscleGroupSecondary.setEnabled(s.length() != 0);

            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        Intent intent = getIntent();
        if (intent != null) {
            if (intent.getSerializableExtra(Helpers.WORKOUTS_KEY) != null) {
                allWorkouts = (ArrayList<Workout>) intent.getSerializableExtra(Helpers.WORKOUTS_KEY);
                position = intent.getIntExtra(Helpers.POSITION_KEY, 0);
                isEdited = true;
                Helpers.setupActionBar(getString(R.string.editing)+" "+ allWorkouts.get(position).getName(),"",getSupportActionBar(),this);
                edtWorkoutName.setText(allWorkouts.get(position).getName());
                edtWorkoutName.setSelection(edtWorkoutName.getText().length());
                edtMuscleGroup.setText(allWorkouts.get(position).getMuscleGroup());
                edtMuscleGroupSecondary.setText((allWorkouts.get(position).getMuscleGroupSecondary()));
                btnAdd.setText(R.string.save);
                switch (allWorkouts.get(position).getType()) {
                    case "FBW":
                        spinnerWorkoutType.setSelection(1);
                        break;
                    case "Push":
                        spinnerWorkoutType.setSelection(2);
                        break;
                    case "Pull":
                        spinnerWorkoutType.setSelection(3);
                        break;
                    case "Legs":
                        spinnerWorkoutType.setSelection(4);
                        break;
                    case "Upper":
                        spinnerWorkoutType.setSelection(5);
                        break;
                    case "Lower":
                        spinnerWorkoutType.setSelection(6);
                        break;
                    case "Split":
                        spinnerWorkoutType.setSelection(7);
                        break;
                    default:
                        spinnerWorkoutType.setSelection(0);
                        break;
                }
            }
        }

        btnAdd.setOnClickListener(v -> {
            if (isEdited) {
                if (!edtWorkoutName.getText().toString().isEmpty() &&
                        !edtWorkoutName.getText().toString().equals(allWorkouts.get(position).getName())) {
                    changeMade = true;
                    allWorkouts.get(position).setName(edtWorkoutName.getText().toString());
                }
                if (!spinnerWorkoutType.getSelectedItem().toString().equals(allWorkouts.get(position).getType())) {
                    changeMade = true;
                    allWorkouts.get(position).setType(spinnerWorkoutType.getSelectedItem().toString());
                }
                if (edtMuscleGroup.getText().toString().isEmpty()) {
                    allWorkouts.get(position).setMuscleGroup("");
                    allWorkouts.get(position).setMuscleGroupSecondary("");
                    changeMade = true;
                } else {
                    if (!edtMuscleGroup.getText().toString().equals(allWorkouts.get(position).getMuscleGroup())) {
                        changeMade = true;
                        allWorkouts.get(position).setMuscleGroup(edtMuscleGroup.getText().toString());
                    }
                    if (!edtMuscleGroupSecondary.getText().toString().equals(allWorkouts.get(position).getMuscleGroupSecondary())) {
                        changeMade = true;
                        allWorkouts.get(position).setMuscleGroupSecondary(edtMuscleGroupSecondary.getText().toString());
                    }
                }
                if(changeMade){
                    Utils.getInstance(AddWorkoutActivity.this).updateWorkouts(allWorkouts);
                }
//                    Intent intent=new Intent(AddWorkoutActivity.this, MainActivity.class);
//                    intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
//                    startActivity(intent);
                finish();


            } else {
                if (edtWorkoutName.getText().toString().isEmpty()) {
                    Toast.makeText(AddWorkoutActivity.this, R.string.nameIsRequired, Toast.LENGTH_SHORT).show();
                    imgWorkoutName.setImageResource(R.drawable.ic_hexagon_single_red);
                    shake(imgWorkoutName);
                } else {
                    String workoutName = edtWorkoutName.getText().toString();
                    String workoutType = spinnerWorkoutType.getSelectedItem().toString();
                    String muscleGroup = edtMuscleGroup.getText().toString();
                    String muscleGroupSecondary = "";
                    if (edtMuscleGroupSecondary.isEnabled()) {
                        muscleGroupSecondary = edtMuscleGroupSecondary.getText().toString();
                    }
                    Workout workout = new Workout(5, workoutName, workoutType, muscleGroup, muscleGroupSecondary, 0);
//                        Toast.makeText(AddWorkoutActivity.this, workout.toString(), Toast.LENGTH_SHORT).show();
                    Intent returnIntent = new Intent();
                    returnIntent.putExtra(Helpers.NEW_WORKOUT_KEY, workout);
                    setResult(Activity.RESULT_OK, returnIntent);
                    InputMethodManager imm1;
                    imm1 = (InputMethodManager) getSystemService(Activity.INPUT_METHOD_SERVICE);
                    imm1.toggleSoftInput(InputMethodManager.HIDE_IMPLICIT_ONLY, 0);
                    finish();
                }
            }

        });
    }

    private void shake(View v) {
        ObjectAnimator
                .ofFloat(v, "translationX", 0, 25, -25, 25, -25, 15, -15, 6, -6, 0)
                .setDuration(200)
                .start();
    }
}

File: app/src/main/java/com/example/gymbuddy/view/AddWorkoutActivity.kt
package com.example.gymbuddy.view

import android.animation.ObjectAnimator
import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import com.example.gymbuddy.R
import com.example.gymbuddy.databinding.ActivityAddWorkoutBinding
import com.example.gymbuddy.helpers.Helpers
import com.example.gymbuddy.helpers.Utils
import com.example.gymbuddy.model.Workout
import java.util.*
import kotlin.collections.ArrayList

class AddWorkoutActivity : AppCompatActivity() {
    private var isEdited = false
    private var changeMade = false
    private lateinit var allWorkouts: ArrayList<Workout>
    private var position = 0
    private lateinit var binding: ActivityAddWorkoutBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        setTheme(R.style.Theme_GymBuddy)
        super.onCreate(savedInstanceState)
        binding = ActivityAddWorkoutBinding.inflate(layoutInflater)
        setContentView(binding.root)
        Helpers.setupActionBar(getString(R.string.addNewWorkout), "", supportActionBar, this)
        binding.edtWorkoutName.requestFocus()
        val imm = getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
        imm.toggleSoftInput(InputMethodManager.SHOW_FORCED, 0)
        binding.tilMuscleGroupSecondary.isEnabled = false
        binding.edtMuscleGroup.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                binding.tilMuscleGroupSecondary.isEnabled = s.isNotEmpty()
            }

            override fun afterTextChanged(s: Editable) {}
        })
        if (intent != null) {
            if (intent.getSerializableExtra(Helpers.WORKOUTS_KEY) != null) {
                allWorkouts = intent.getSerializableExtra(Helpers.WORKOUTS_KEY) as ArrayList<Workout>? ?: ArrayList()
                position = intent.getIntExtra(Helpers.POSITION_KEY, 0)
                isEdited = true
                Helpers.setupActionBar(getString(R.string.editing) + " " + allWorkouts[position].name, "", supportActionBar, this)
                binding.edtWorkoutName.setText(allWorkouts[position].name)
                binding.edtWorkoutName.setSelection(binding.edtWorkoutName.text?.length ?: 0)
                binding.edtMuscleGroup.setText(allWorkouts[position].muscleGroup)
                binding.edtMuscleGroupSecondary.setText(allWorkouts[position].muscleGroupSecondary)
                binding.btnAdd.setText(R.string.save)
                binding.edtWorkoutType.setText(allWorkouts[position].type, false)
            }else{
                binding.edtWorkoutType.setText("---", false)
            }
        }
        val arrayAdapter = ArrayAdapter(this, R.layout.dropdown_item, resources.getStringArray(R.array.workoutTypes))
        binding.edtWorkoutType.setAdapter(arrayAdapter)
        binding.btnAdd.setOnClickListener {
            if (isEdited) {
                if (binding.edtWorkoutName.text.toString().isNotEmpty() &&
                    binding.edtWorkoutName.text.toString() != allWorkouts[position].name
                ) {
                    changeMade = true
                    allWorkouts[position].name = binding.edtWorkoutName.text.toString()
                }
                if (binding.edtWorkoutType.text.toString() != allWorkouts[position].type) {
                    changeMade = true
                    allWorkouts[position].type = binding.edtWorkoutType.text.toString()
                }
                if (binding.edtMuscleGroup.text.toString().isEmpty()) {
                    allWorkouts[position].muscleGroup = ""
                    allWorkouts[position].muscleGroupSecondary = ""
                    changeMade = true
                } else {
                    if (binding.edtMuscleGroup.text.toString() != allWorkouts[position].muscleGroup) {
                        changeMade = true
                        allWorkouts[position].muscleGroup = binding.edtMuscleGroup.text.toString()
                    }
                    if (binding.edtMuscleGroupSecondary.text.toString() != allWorkouts[position].muscleGroupSecondary) {
                        changeMade = true
                        allWorkouts[position].muscleGroupSecondary = binding.edtMuscleGroupSecondary.text.toString()
                    }
                }
                if (changeMade) {
                    Utils.getInstance(this@AddWorkoutActivity).updateWorkouts(allWorkouts)
                }
                finish()
            } else {
                if (binding.edtWorkoutName.text.toString().isEmpty()) {
                    Toast.makeText(this@AddWorkoutActivity, R.string.nameIsRequired, Toast.LENGTH_SHORT).show()
                    binding.imgWorkoutName.setImageResource(R.drawable.ic_hexagon_single_red)
                    shake(binding.imgWorkoutName)
                } else {
                    val workoutName = binding.edtWorkoutName.text.toString()
                    val workoutType = binding.edtWorkoutType.text.toString()
                    val muscleGroup = binding.edtMuscleGroup.text.toString()
                    var muscleGroupSecondary = ""
                    if (binding.edtMuscleGroupSecondary.isEnabled) {
                        muscleGroupSecondary = binding.edtMuscleGroupSecondary.text.toString()
                    }
                    val workout = Workout(5, workoutName, workoutType, muscleGroup, muscleGroupSecondary, 0)
                    //                        Toast.makeText(AddWorkoutActivity.this, workout.toString(), Toast.LENGTH_SHORT).show();
                    val returnIntent = Intent()
                    returnIntent.putExtra(Helpers.NEW_WORKOUT_KEY, workout)
                    setResult(RESULT_OK, returnIntent)
                    val imm1: InputMethodManager = getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
                    imm1.toggleSoftInput(InputMethodManager.HIDE_IMPLICIT_ONLY, 0)
                    finish()
                }
            }
        }
    }

    private fun shake(v: View?) {
        ObjectAnimator
            .ofFloat(v, "translationX", 0f, 25f, -25f, 25f, -25f, 15f, -15f, 6f, -6f, 0f)
            .setDuration(200)
            .start()
    }
}